pkg/instance/service.go

```diff
@@ -58,6 +61,60 @@ type service struct {
 	helmfileService    helmfile
 }
 
+func (s service) SaveDeployment(deployment *model.Deployment) error {
+	return s.instanceRepository.SaveDeployment(deployment)
+}
+
+func (s service) FindDeploymentById(id uint) (*model.Deployment, error) {
+	return s.instanceRepository.FindDeploymentById(id)
+}
+
+func (s service) SaveInstance(instance *model.DeploymentInstance) error {
+	deployment, err := s.instanceRepository.FindDeploymentById(instance.DeploymentID)
+	if err != nil {
+		return err
+	}
+
+	deployment.Instances = append(deployment.Instances, instance)
+	err = s.validateDeployment(deployment)
+	if err != nil {
+		return err
+	}
+
+	return s.instanceRepository.SaveInstance(instance)
+}
+
+func (s service) validateDeployment(deployment *model.Deployment) error {
+	stacks, err := s.collectStacks(deployment)
+	if err != nil {
+		return err
+	}
+
+	err = stack.ValidateNoCycles(stacks)
+	if err != nil {
+		return err
+	}
+
+	err = stack.ValidateConsumedParameters(stacks)
+	if err != nil {
+		return err
```

Comment:
at least

```suggestion
		return fmt.Errorf("error saving instance for %q as it is missing a required stack: %v", instance.StackName, err)
```

Ideally we would tell them which stack they are missing